File: app/src/main/java/com/boarbeard/io/ExternalMedia.java
package com.boarbeard.io;

import java.io.File;
import java.io.FileOutputStream;
import java.io.InputStream;
import java.util.ArrayList;
import java.util.List;

import android.content.Context;
import android.net.Uri;
import android.os.Environment;

import com.boarbeard.R;

public class ExternalMedia {

	public static void init(Context context) {
		try {
			File dir = new File(Environment.getExternalStorageDirectory()
					.getAbsolutePath() + "/com.boarbeard");
			if (!dir.exists()) {
				dir.mkdirs();

				// Write read me
				InputStream src = context.getResources().openRawResource(
						R.raw.readme);
				FileOutputStream dst = new FileOutputStream(new File(
						dir.getAbsolutePath() + "/readme.txt"));
				byte[] buffer = new byte[src.available()];
				src.read(buffer);
				dst.write(buffer);
				dst.close();
				src.close();

				// Write grammar.xml example
				src = context.getResources().openRawResource(R.raw.grammar);
				dst = new FileOutputStream(new File(dir.getAbsolutePath()
						+ "/grammar.xml"));
				buffer = new byte[src.available()];
				src.read(buffer);
				dst.write(buffer);
				src.close();
				dst.close();
			}
		} catch (Exception e) {
		}
	}

	public static boolean isReadable() {
		boolean mExternalStorageAvailable = false;
		String state = Environment.getExternalStorageState();

		if (Environment.MEDIA_MOUNTED.equals(state)) {
			// We can read and write the media
			mExternalStorageAvailable = true;
		} else if (Environment.MEDIA_MOUNTED_READ_ONLY.equals(state)) {
			// We can only read the media
			mExternalStorageAvailable = true;
		} else {
			// Something else is wrong. It may be one of many other states, but
			// all we need
			// to know is we can neither read nor write
			mExternalStorageAvailable = false;
		}

		return mExternalStorageAvailable;
	}

	public static List<Uri> getMediaFolders(Context context) {
		List<Uri> folders = new ArrayList<Uri>();

		if (ExternalMedia.isReadable()) {
			try {
				File dir = new File(Environment.getExternalStorageDirectory()
						.getAbsolutePath() + "/com.boarbeard");
				for (File file : dir.listFiles()) {
					if (file.isDirectory())
						folders.add(Uri.fromFile(file));
				}
			} catch (Exception e) {
			}
		}

		return folders;
	}

	public static List<Uri> getMediaFiles(File directory) {
		List<Uri> medias = new ArrayList<Uri>();
		if (ExternalMedia.isReadable()) {
			try {
				for (File file : directory.listFiles()) {
					medias.add(Uri.fromFile(file));
				}
			} catch (Exception e) {

			}
		}

		return medias;
	}

	public static File getMediaTextFile(String folder) {
		if (ExternalMedia.isReadable()) {
			try {
				File file = new File(Environment.getExternalStorageDirectory()
						.getAbsolutePath()
						+ "/com.boarbeard/"
						+ folder
						+ "/grammar.xml");
				if (file.isFile())
					return file;
			} catch (Exception e) {
			}
		}

		return null;
	}
}

File: app/src/main/java/com/boarbeard/io/ExternalMedia.kt
package com.boarbeard.io

import android.content.Context
import android.net.Uri
import android.os.Environment
import com.boarbeard.R
import java.io.File
import java.io.FileOutputStream
import java.io.IOException

object ExternalMedia {
    lateinit var media_context: Context

    @JvmStatic
    fun init(context: Context) {
        media_context = context

        if (isExternalStorageWritable() && !File(context.getExternalFilesDir(null), "readme.txt").exists()) {
            // Write read me
            try {
                context.resources.openRawResource(R.raw.readme).use { input ->
                    FileOutputStream(File(context.getExternalFilesDir(null), "readme.xml")).use { output ->
                        input.buffered().copyTo(out = output)
                    }
                }
            } catch (e: IOException) {
                e.printStackTrace()
            }

            // write grammar.xml
            try {
                context.resources.openRawResource(R.raw.grammar).use { input ->
                    FileOutputStream(File(context.getExternalFilesDir(null), "grammar.xml")).use { output ->
                        input.buffered().copyTo(out = output)
                    }
                }
            } catch (e: IOException) {
                e.printStackTrace()
            }
        }

    }

    // Checks if a volume containing external storage is available
    // for read and write.
    private fun isExternalStorageWritable(): Boolean {
        return Environment.getExternalStorageState() == Environment.MEDIA_MOUNTED
    }

    // Checks if a volume containing external storage is available to at least read.
    private fun isExternalStorageReadable(): Boolean {
        return Environment.getExternalStorageState() in
                setOf(Environment.MEDIA_MOUNTED, Environment.MEDIA_MOUNTED_READ_ONLY)
    }

    @JvmStatic
    fun getMediaFolders(context: Context?): List<Uri> {
        if (isExternalStorageReadable()) {
            try {
                return context?.getExternalFilesDir(null)?.listFiles()?.map { file ->
                    Uri.fromFile(file)
                } ?: emptyList()
            } catch (e: Exception) {
                e.printStackTrace()
            }
        }
        return emptyList()
    }

    @JvmStatic
    fun getMediaTextFile(folder: String): File? {
        if (isExternalStorageReadable()) {
            try {
                val file = File(File(media_context.getExternalFilesDir(null), folder), "grammar.xml")
                if (file.isFile) return file
            } catch (e: Exception) {
                e.printStackTrace()
            }
        }

        return null
    }
}